cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,426 @@
+ * @tparam mask_t
+ */
+template <typename vertex_t, typename edge_t, typename mask_t = std::uint32_t>
+struct graph_mask_t {
+ public:
+  using vertex_type = vertex_t;
+  using edge_type   = edge_t;
+  using mask_type   = mask_t;
+  using size_type   = std::size_t;
+
+  ~graph_mask_t()                       = default;
+  graph_mask_t(graph_mask_t&&) noexcept = default;
+
+  graph_mask_t() = delete;
+
+  explicit graph_mask_t(raft::handle_t const& handle,
+                        vertex_t n_vertices,
+                        edge_t n_edges,
+                        bool complement = false)
+    : handle_(handle),
+      n_vertices_(n_vertices),
+      n_edges_(n_edges),
+      edges_(0, handle.get_stream()),
+      vertices_(0, handle.get_stream()),
+      complement_(complement)
+  {
+  }
+
+  explicit graph_mask_t(graph_mask_t const& other)
+    : handle_(other.handle_),
+      n_vertices_(other.n_vertices_),
+      n_edges_(other.n_edges_),
+      edges_(other.edges_, other.handle_.get_stream()),
+      vertices_(other.vertices_, other.handle_.get_stream()),
+      complement_(other.complement_)
+  {
+  }
+
+  graph_mask_t& operator=(graph_mask_t&&) noexcept = default;
+  graph_mask_t& operator=(graph_mask_t const& other) = default;
+
+  /**
+   * Determines whether the 1 bit in a vertex or edge position
+   * represents an inclusive mask or exclusive mask. Default is
+   * an inclusive mask (e.g. 1 bit means the corresponding vertex
+   * or edge should be included in computations).
+   * @return
+   */
+  bool is_complemented() const { return complement_; }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with an edge mask.
+   * @return
+   */
+  bool has_edge_mask() const { return get_edge_mask().has_value(); }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with a vertex mask.
+   * @return
+   */
+  bool has_vertex_mask() const { return get_vertex_mask().has_value(); }
+
+  /**
+   * Returns the edge mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_edge_mask() const
+  {
+    return edges_.size() > 0 ? std::make_optional<mask_t const*>(edges_.data()) : std::nullopt;
+  }
+
+  /**
+   * Retuns the vertex mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_vertex_mask() const
+  {
+    return vertices_.size() > 0 ? std::make_optional<mask_t const*>(vertices_.data())
+                                : std::nullopt;
+  }
+
+  vertex_t get_n_vertices() { return n_vertices_; }
+
+  edge_t get_n_edges() { return n_edges_; }
+
+  /**
+   * Initializes an edge mask by allocating the device memory
+   */
+  void initialize_edge_mask()
+  {
+    if (edges_.size() == 0) {
+      edges_.resize(n_edges_, handle_.get_stream());
+      clear_edge_mask();
+    }
+  }
+
+  /**
+   * Initializes a vertex mask by allocating the device memory
+   */
+  void initialize_vertex_mask()
```

Comment:
"initialize" might be interpreted as filling mask values to some initial values. And should we have a separate function for allocation/initialization or can't we allocate & initialize in a constructor?

I think graph_view_t can have a member function to return a graph_mask_t object.

Reply:
I don't mind having separate methods for `allocate()` and `initialize()` if that makes it easier to use. Rather than different constructors, I'd prefer if we use a either static factory methods or builders to construct and populate a mask, for example when it is known at construction time that an edge mask and/or vertex mask will be needed.

Thinking out loud, I'm kind of partial to something like this very rough idea:

```c++
graph_mask_t<vertex_t, edge_t, mask_t> graph_mask_builder_t(handle, n_vertices, n_edges)
        .edge_mask(e_mask_func mask_function, bool init = 0)   // init will first do a thrust::fill or cudamemset
        .vertex_mask(v_mask_func mask_function, bool init = 0)  // init will first do a thrust::fill or cudamemset
        .build();
```

This way, the underlying mask owning object can stay mutable even wrt its memory allocations but all the internals and complexities can remain hidden behind the builder API. This also allows a user to build a mask object with edge and/or vertex  masks without having to maintain that logic in multiple constructors (nor make a single constructor more complex than it needs to be).